src/ballet/block/fd_microblock.c

```diff
@@ -0,0 +1,181 @@
+#include "fd_microblock.h"
+
+#include "../bmtree/fd_bmtree.h"
+
+FD_FN_CONST ulong
+fd_microblock_align( void ) {
+  return FD_MICROBLOCK_ALIGN;
+}
+
+FD_FN_CONST ulong
+fd_microblock_footprint( ulong txn_cnt ) {
+  /* Fixed-size data preceding VLA */
+  ulong hdr_sz = sizeof(fd_microblock_t);
+
+  /* Max required heap space for all txn descriptors */
+  if( FD_UNLIKELY( txn_cnt > (ULONG_MAX-hdr_sz)/FD_TXN_MAX_SZ ) ) return 0; /* txn_cnt too large */
+  ulong txn_tbl_sz = txn_cnt*FD_TXN_MAX_SZ;
+
+  return fd_ulong_align_up( hdr_sz + txn_tbl_sz, FD_MICROBLOCK_ALIGN );
+}
+
+void *
+fd_microblock_new( void * shmem,
+                   ulong  txn_cnt ) {
+
+  if( FD_UNLIKELY( !shmem ) ) {
+    FD_LOG_WARNING(( "NULL shmem" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)shmem, fd_microblock_align() ) ) ) {
+    FD_LOG_WARNING(( "misaligned shmem" ));
+    return NULL;
+  }
+
+  ulong footprint = fd_microblock_footprint( txn_cnt );
+  if( FD_UNLIKELY( !footprint ) ) {
+    FD_LOG_WARNING(( "bad txn_cnt (%lu)", txn_cnt ));
+    return NULL;
+  }
+
+  fd_memset( shmem, 0, footprint );
+
+  fd_microblock_t * hdr = (fd_microblock_t *)shmem;
+  hdr->txn_max_cnt = txn_cnt;
+
+  FD_COMPILER_MFENCE();
+  hdr->magic = FD_MICROBLOCK_MAGIC;
+  FD_COMPILER_MFENCE();
+
+  return shmem;
+}
+
+fd_microblock_t *
+fd_microblock_join( void * shblock ) {
+
+  if( FD_UNLIKELY( !shblock ) ) {
+    FD_LOG_WARNING(( "NULL shblock" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)shblock, fd_microblock_align() ) ) ) {
+    FD_LOG_WARNING(( "misaligned shblock" ));
+    return NULL;
+  }
+
+  fd_microblock_t * hdr = (fd_microblock_t *)shblock;
+  if( FD_UNLIKELY( hdr->magic!=FD_MICROBLOCK_MAGIC ) ) {
+    FD_LOG_WARNING(( "bad magic" ));
+    return NULL;
+  }
+
+  return (fd_microblock_t *)shblock;
+}
+
+void *
+fd_microblock_leave( fd_microblock_t * block ) {
+
+  if( FD_UNLIKELY( !block ) ) {
+    FD_LOG_WARNING(( "NULL block" ));
+    return NULL;
+  }
+
+  return (void *)block;
+}
+
+void *
+fd_microblock_delete( void * shblock ) {
+
+  if( FD_UNLIKELY( !shblock ) ) {
+    FD_LOG_WARNING(( "NULL shblock" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)shblock, fd_microblock_align() ) ) ) {
+    FD_LOG_WARNING(( "misaligned shblock" ));
+    return NULL;
+  }
+
+  fd_microblock_t * hdr = (fd_microblock_t *)shblock;
+  if( FD_UNLIKELY( hdr->magic!=FD_MICROBLOCK_MAGIC ) ) {
+    FD_LOG_WARNING(( "bad magic" ));
+    return NULL;
+  }
+
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( hdr->magic ) = 0UL;
+  FD_COMPILER_MFENCE();
+
+  return shblock;
+}
+
+int
+fd_microblock_deserialize( fd_microblock_t * block,
+                           void **           buf,
+                           ulong *           buf_sz,
+                           fd_txn_parse_counters_t * counters_opt ) {
+# define ADVANCE(n) do {                       \
```

Comment:
Just a formatting nitpick, but I think the whole block should be indented and the `do` should go on its own line.
```suggestion
# define ADVANCE(n)                            \
  do {                                         \
```